File: JNLuaEngine/src/main/java/org/anime_game_servers/jnlua_engine/JNLuaReflector.java
package org.anime_game_servers.jnlua_engine;

import io.github.oshai.kotlinlogging.KLogger;
import io.github.oshai.kotlinlogging.KotlinLogging;
import lombok.Getter;
import org.anime_game_servers.lua.utils.LuaHelpersJvmKt;
import org.terasology.jnlua.*;

import java.beans.BeanInfo;
import java.beans.IntrospectionException;
import java.beans.Introspector;
import java.beans.PropertyDescriptor;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.*;

public class JNLuaReflector extends DefaultJavaReflector {
    private static KLogger logger = KotlinLogging.INSTANCE.logger(JNLuaReflector.class.getName());
    @Getter
    private static final JNLuaReflector INSTANCE = new JNLuaReflector();

    @Override
    protected Map<String, Accessor> createClassAccessors(Class<?> clazz) {
        Map<String, Accessor> result = new HashMap<String, Accessor>();

        // Fields
        Field[] fields = clazz.getFields();
        for (int i = 0; i < fields.length; i++) {
            var field = fields[i];
            var luaNames = LuaHelpersJvmKt.getLuaNames(field);
            for (var luaName : luaNames) {
                result.put(luaName, new FieldAccessor(field));
            }
        }

        // Methods
        Map<String, Map<List<Class<?>>, Invocable>> accessibleMethods = new HashMap<String, Map<List<Class<?>>, Invocable>>();
        Method[] methods = clazz.getMethods();
        for (int i = 0; i < methods.length; i++) {
            // Do not overwrite fields
            Method method = methods[i];
            if (result.containsKey(method.getName())) {
                continue;
            }

            // Attempt to find the method in a public class if the declaring
            // class is not public
            if (!Modifier.isPublic(method.getDeclaringClass().getModifiers())) {
                method = getPublicClassMethod(clazz, method.getName(),
                        method.getParameterTypes());
                if (method == null) {
                    continue;
                }
            }

            // For each method name and parameter type list, keep
            // only the method declared by the most specific class
            Map<List<Class<?>>, Invocable> overloaded = accessibleMethods
                    .get(method.getName());
            if (overloaded == null) {
                overloaded = new HashMap<List<Class<?>>, Invocable>();
                accessibleMethods.put(method.getName(), overloaded);
            }
            List<Class<?>> parameterTypes = Arrays.asList(method
                    .getParameterTypes());
            Invocable currentInvocable = overloaded.get(parameterTypes);
            if (currentInvocable != null
                    && method.getDeclaringClass().isAssignableFrom(
                    currentInvocable.getDeclaringClass())) {
                continue;
            }
            overloaded.put(parameterTypes, new InvocableMethod(method));
        }
        for (Map.Entry<String, Map<List<Class<?>>, Invocable>> entry : accessibleMethods
                .entrySet()) {
            result.put(entry.getKey(), new InvocableAccessor(clazz, entry
                    .getValue().values()));
        }

        // Constructors
        Constructor<?>[] constructors = clazz.getConstructors();
        List<Invocable> accessibleConstructors = new ArrayList<Invocable>(
                constructors.length);
        for (int i = 0; i < constructors.length; i++) {
            // Ignore constructor if the declaring class is not public
            if (!Modifier.isPublic(constructors[i].getDeclaringClass()
                    .getModifiers())) {
                continue;
            }
            accessibleConstructors
                    .add(new InvocableConstructor(constructors[i]));
        }
        if (clazz.isInterface()) {
            accessibleConstructors.add(new InvocableProxy(clazz));
        }
        if (!accessibleConstructors.isEmpty()) {
            result.put("new", new InvocableAccessor(clazz,
                    accessibleConstructors));
        }

        // Properties
        BeanInfo beanInfo;
        try {
            beanInfo = Introspector.getBeanInfo(clazz);
        } catch (IntrospectionException e) {
            throw new RuntimeException(e);
        }
        PropertyDescriptor[] propertyDescriptors = beanInfo
                .getPropertyDescriptors();
        for (int i = 0; i < propertyDescriptors.length; i++) {
            // Do not overwrite fields or methods
            if (result.containsKey(propertyDescriptors[i].getName())) {
                continue;
            }

            // Attempt to find the read/write methods in a public class if the
            // declaring class is not public
            Method method = propertyDescriptors[i].getReadMethod();
            if (method != null
                    && !Modifier.isPublic(method.getDeclaringClass()
                    .getModifiers())) {
                method = getPublicClassMethod(clazz, method.getName(),
                        method.getParameterTypes());
                try {
                    propertyDescriptors[i].setReadMethod(method);
                } catch (IntrospectionException e) {
                }
            }
            method = propertyDescriptors[i].getWriteMethod();
            if (method != null
                    && !Modifier.isPublic(method.getDeclaringClass()
                    .getModifiers())) {
                method = getPublicClassMethod(clazz, method.getName(),
                        method.getParameterTypes());
                try {
                    propertyDescriptors[i].setWriteMethod(method);
                } catch (IntrospectionException e) {
                }
            }

            // Do not process properties without a read and a write method
            if (propertyDescriptors[i].getReadMethod() == null
                    && propertyDescriptors[i].getWriteMethod() == null) {
                continue;
            }
            result.put(propertyDescriptors[i].getName(), new PropertyAccessor(
                    clazz, propertyDescriptors[i]));
        }
        return result;
    }
}

File: JNLuaEngine/src/main/java/org/anime_game_servers/jnlua_engine/JNLuaReflector.kt
package org.anime_game_servers.jnlua_engine

import io.github.oshai.kotlinlogging.KotlinLogging.logger
import org.anime_game_servers.lua.utils.getLuaNames
import org.terasology.jnlua.DefaultJavaReflector
import java.beans.BeanInfo
import java.beans.IntrospectionException
import java.beans.Introspector
import java.lang.reflect.Modifier
import java.util.*
import kotlin.reflect.jvm.kotlinProperty

private val logger = logger {}

object JNLuaReflector : DefaultJavaReflector() {
    override fun createClassAccessors(clazz: Class<*>): Map<String, Accessor> {
        val result: MutableMap<String, Accessor> = HashMap<String, Accessor>()

        // Fields
        val fields = clazz.getFields()
        for (i in fields.indices) {
            val field = fields[i]
            val luaNames = field.getLuaNames()
            for (luaName in luaNames) {
                result[luaName] = FieldAccessor(field)
            }
        }

        // Methods
        val accessibleMethods: MutableMap<String, MutableMap<List<Class<*>>, Invocable>> =
            HashMap<String, MutableMap<List<Class<*>>, Invocable>>()
        val methods = clazz.getMethods()
        for (i in methods.indices) {
            // Do not overwrite fields
            var method = methods[i]
            if (result.containsKey(method.name)) {
                continue
            }

            // Attempt to find the method in a public class if the declaring
            // class is not public
            if (!Modifier.isPublic(method.declaringClass.modifiers)) {
                method = getPublicClassMethod(
                    clazz, method.name,
                    method.parameterTypes
                )
                if (method == null) {
                    continue
                }
            }

            // For each method name and parameter type list, keep
            // only the method declared by the most specific class
            var overloaded = accessibleMethods[method.name]
            if (overloaded == null) {
                overloaded = HashMap<List<Class<*>>, Invocable>()
                accessibleMethods[method.name] = overloaded
            }
            val parameterTypes = mutableListOf(*method.parameterTypes)
            val currentInvocable = overloaded[parameterTypes]
            if (currentInvocable != null
                && method.declaringClass.isAssignableFrom(
                    currentInvocable.declaringClass
                )
            ) {
                continue
            }
            overloaded[parameterTypes] = InvocableMethod(method)
        }
        for (entry in accessibleMethods.entries) {
            result[entry.key] = InvocableAccessor(clazz, entry.value.values)
        }

        // Constructors
        val constructors = clazz.getConstructors()
        val accessibleConstructors: MutableList<Invocable> = ArrayList<Invocable>(
            constructors.size
        )
        for (i in constructors.indices) {
            // Ignore constructor if the declaring class is not public
            if (!Modifier.isPublic(
                    constructors[i].getDeclaringClass().modifiers
                )
            ) {
                continue
            }
            accessibleConstructors
                .add(InvocableConstructor(constructors[i]))
        }
        if (clazz.isInterface) {
            accessibleConstructors.add(InvocableProxy(clazz))
        }
        if (!accessibleConstructors.isEmpty()) {
            result["new"] = InvocableAccessor(
                clazz,
                accessibleConstructors
            )
        }

        // Properties
        val beanInfo: BeanInfo
        try {
            beanInfo = Introspector.getBeanInfo(clazz)
        } catch (e: IntrospectionException) {
            throw RuntimeException(e)
        }
        val propertyDescriptors = beanInfo.propertyDescriptors
        for (propertyDescriptor in propertyDescriptors) {
            // Do not overwrite fields or methods
            if (result.containsKey(propertyDescriptor.name)) {
                continue
            }

            // Attempt to find the read/write methods in a public class if the
            // declaring class is not public
            var method = propertyDescriptor.getReadMethod()
            if (method != null
                && !Modifier.isPublic(
                    method.declaringClass.modifiers
                )
            ) {
                method = getPublicClassMethod(
                    clazz, method.name,
                    method.parameterTypes
                )
                try {
                    propertyDescriptor.setReadMethod(method)
                } catch (e: IntrospectionException) {
                }
            }
            method = propertyDescriptor.getWriteMethod()
            if (method != null
                && !Modifier.isPublic(
                    method.declaringClass.modifiers
                )
            ) {
                method = getPublicClassMethod(
                    clazz, method.name,
                    method.parameterTypes
                )
                try {
                    propertyDescriptor.setWriteMethod(method)
                } catch (e: IntrospectionException) {
                }
            }

            // Do not process properties without either a read or a write method
            if (propertyDescriptor.getReadMethod() == null
                && propertyDescriptor.getWriteMethod() == null
            ) {
                continue
            }
            val name = propertyDescriptor.name
            val accessor = PropertyAccessor(clazz, propertyDescriptor)
            try {
                val declaredField = clazz.getDeclaredField(name)
                val ktProperty = declaredField.kotlinProperty
                if (ktProperty != null) {
                    val luaNames: List<String> = ktProperty.annotations.getLuaNames(name)
                    for (luaName in luaNames) {
                        result[luaName] = accessor
                    }
                }
            } catch (e: NoSuchFieldException) {
                logger.warn(e) { "[createClassAccessors] " }
            }

            result[name] = accessor
        }
        return result
    }
}
